File: dart/test/pi_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:layrz_protocol/layrz_protocol.dart';

void main() {
  test('PiPacket.parse()', () {
    String ident = 'testident';
    String firmwareId = '1';
    int firmwareBuild = 1;
    String deviceId = '1';
    String hardwareId = '1';
    String modelId = '1';
    FirmwareBranch firmwareBranch = FirmwareBranch.development;
    bool fotaEnabled = true;

    String payload = '$ident;';
    payload += '$firmwareId;';
    payload += '$firmwareBuild;';
    payload += '$deviceId;';
    payload += '$hardwareId;';
    payload += '$modelId;';
    payload += '${firmwareBranch.toJson()};';
    // payload += '${fotaEnabled ? '1' : '0'};';
    payload += '1;';

    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();

    payload = '<Pi>$payload$crc</Pi>';

    PiPacket link = PiPacket.fromPacket(payload);

    expect(link.ident, ident);
    expect(link.firmwareId, firmwareId);
    expect(link.firmwareBuild, firmwareBuild);
    expect(link.deviceId, deviceId);
    expect(link.hardwareId, hardwareId);
    expect(link.modelId, modelId);
    expect(link.firmwareBranch, firmwareBranch);
    expect(link.fotaEnabled, fotaEnabled);

    String reversedPayload = link.toPacket();
    expect(reversedPayload, payload);
  });
}

File: go/packets/client/pi_test.go
package client_test

import (
	"testing"

	"github.com/goldenm-software/layrz-protocol/go/v3/definitions"
	"github.com/goldenm-software/layrz-protocol/go/v3/packets/client"
)

func TestPi_FromPacket_ToPacket(t *testing.T) {
	tests := []struct {
		name   string
		ident  string
		branch definitions.FirmwareBranch
	}{
		{
			name:   "stable firmware",
			ident:  "test_device",
			branch: definitions.Stable,
		},
		{
			name:   "development firmware",
			ident:  "dev_device",
			branch: definitions.Development,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			packet := client.PiPacket{
				Ident:          tt.ident,
				FirmwareId:     "fw1",
				FirmwareBuild:  1,
				DeviceId:       1,
				HardwareId:     1,
				ModelId:        1,
				FirmwareBranch: tt.branch,
				FotaEnabled:    false,
			}
			encoded := *packet.ToPacket()

			raw := encoded
			decoded := client.PiPacket{}
			if err := decoded.FromPacket(&raw); err != nil {
				t.Fatalf("FromPacket failed: %v", err)
			}
			if decoded.FirmwareBranch != tt.branch {
				t.Errorf("branch mismatch: got %s, want %s", decoded.FirmwareBranch, tt.branch)
			}
			if *decoded.ToPacket() != encoded {
				t.Errorf("round-trip mismatch")
			}
		})
	}
}
